### PR TITLE
SDK-418 [Java SDK] SSE mode support

### PR DESCRIPTION
**Pull request by VladimirLevitskiy**
Work relates to:
----------------
Link(s) to Jira Task/User Story/Bug/Epic (if any):
- [SDK-418](https://incountry.atlassian.net/browse/SDK-418)

Code Quality (Before submitting PR)
------------
- [x] Ensure code compiles and passes Sonar Lint.
- [x] Do your best to ensure integration tests pass. If they don't then...
- [x] Remove, ignore, or mark pending any brittle tests.
- [x] Remove commented out code.

Target Release
--------------
- [ ] Please put the target release vehicle this needs to be bundled with.
- [ ] Did you increment the version? What is the new version?

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -107,13 +107,18 @@ public class StorageConfig {
    private String clientSecret;
    /** Optional. Defines custom API URL, can also be set via INC_ENDPOINT */
    private String endPoint;
-   /** Instance of SecretKeyAccessor class. Used to fetch encryption secret */
+   /** Instance of SecretKeyAccessor class. Used to fetch encryption secret 
+    *  If null - data will be stored without encryption as Base64 or server side can be used*/
```
Comment (RommeroInCountry) [inline README.md:111]: If null - data will be stored without encryption in case server side encryption mode is also disabled

File: src/main/java/com/incountry/residence/sdk/StorageConfig.java [hunk 2]
```diff
@@ -322,6 +327,9 @@ public boolean isHashSearchKeys() {
      */
     public StorageConfig setHashSearchKeys(boolean hashSearchKeys) {
         this.hashSearchKeys = hashSearchKeys;
+        if (hashSearchKeys) {
+            this.serverSideEncryption = false;
+        }
```
Comment (RommeroInCountry) [inline src/main/java/com/incountry/residence/sdk/StorageConfig.java:332]: let's discuss this one on the call

File: src/main/java/com/incountry/residence/sdk/StorageConfig.java [hunk 3]
```diff
@@ -385,11 +393,32 @@ public StorageConfig setRetryMaxDelay(Integer retryMaxDelay) {
         return this;
     }
 
+    public boolean isServerSideEncryption() {
+        return serverSideEncryption;
+    }
+
+    /**
+     * Optional. Default is false. Enables a server side encryption mode. Unencrypted data will be sent to storage via HTTPS,
```
Comment (RommeroInCountry) [inline src/main/java/com/incountry/residence/sdk/StorageConfig.java:401]: See README comment

File: src/main/java/com/incountry/residence/sdk/tools/DtoTransformer.java [hunk 4]
```diff
@@ -79,44 +82,71 @@ public SecretKeyAccessor getKeyAccessor() {
...
     public TransferRecord getTransferRecord(Record originalRecord) throws StorageClientException, StorageCryptoException {
-        TransferRecord meta = cloneRecordForMeta(originalRecord);
-        ComplexBody newBodyObject = new ComplexBody(meta, originalRecord.getBody());
+        TransferRecord tempRecord = cloneRecordStringFields(originalRecord);
+        if (noTransformations) {
+            tempRecord.setBody(originalRecord.getBody());
+            tempRecord.setPrecommitBody(originalRecord.getPrecommitBody());
+            fillCommonFields(tempRecord, originalRecord);
+            return tempRecord;
+        }
```
Comment (RommeroInCountry) [inline src/main/java/com/incountry/residence/sdk/tools/DtoTransformer.java:91]: Let's remove this code optimization in favour of readability. Otherwise it's hard to understand why copy strings first and then copy rangeKeys + date keys.

File: src/main/java/com/incountry/residence/sdk/tools/DtoTransformer.java [hunk 5]
```diff
@@ -62,12 +62,15 @@
...
+    private final boolean noTransformations;
 
-    public DtoTransformer(CryptoProvider cryptoProvider, HashUtils hashUtils, boolean hashSearchKeys, SecretKeyAccessor keyAccessor) {
+    public DtoTransformer(CryptoProvider cryptoProvider, HashUtils hashUtils, boolean hashSearchKeys,
+                          SecretKeyAccessor keyAccessor, boolean noTransformations) {
         this.cryptoProvider = cryptoProvider;
         this.hashUtils = hashUtils;
         this.hashSearchKeys = hashSearchKeys;
         this.keyAccessor = keyAccessor;
+        this.noTransformations = noTransformations;
```
Comment (RommeroInCountry) [inline src/main/java/com/incountry/residence/sdk/tools/DtoTransformer.java:73]: can we just not use DtoTransformer in case of SSE mode? 
Or probs have EncryptedDtoTransformer and SSEDtoTransformer (or BypassDtoTransformer/PlainDtoTransformer) classes?

File: src/test/java/com/incountry/residence/sdk/RecordTest.java [hunk 6]
```diff
@@ -348,12 +349,26 @@ void dateFieldTest() throws StorageClientException, StorageServerException, Stor
...
+        assertTrue(Record.stringEquals(null, null, false));
+        assertFalse(Record.stringEquals(null, "value", false));
+        assertFalse(Record.stringEquals("value", null, false));
+        assertTrue(Record.stringEquals("value", "value", false));
+        assertFalse(Record.stringEquals("value", "Value", false));
+        assertTrue(Record.stringEquals("value", "value", true));
+        assertTrue(Record.stringEquals("value", "Value", true));
+        assertFalse(Record.stringEquals(null, "Value", true));
+        assertTrue(Record.stringEquals(null, null, true));
+        assertFalse(Record.stringEquals("value", "Value1", true));
```
Comment (RommeroInCountry) [inline src/test/java/com/incountry/residence/sdk/RecordTest.java:372]: Let's go with all True first and all False last :D (separated with a newline)

File: src/test/java/com/incountry/residence/sdk/PlainDtoTransformerTest.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    void nullSecretAccessor() {
+        PlainDtoTransformer transformer = new PlainDtoTransformer();
+        assertNull(transformer.getKeyAccessor());
+    }
+
+    @Test
+    void getNullRecord() throws StorageServerException, StorageCryptoException, StorageClientException {
+        PlainDtoTransformer transformer = new PlainDtoTransformer();
+        assertNull(transformer.getRecord(null));
+    }
```
Comment (RommeroInCountry) [inline src/test/java/com/incountry/residence/sdk/PlainDtoTransformerTest.java:23]: let's add getRecord test with non-null record :)

File: src/main/java/com/incountry/residence/sdk/tools/transformer/EncryptedDtoTransformer.java [hunk 8]
```diff
@@ -35,88 +25,103 @@
...
-                .setKey18(sourceRecord.getKey18())
-                .setKey19(sourceRecord.getKey19())
-                .setKey20(sourceRecord.getKey20())
-                .setParentKey(sourceRecord.getParentKey())
-                .setServiceKey1(sourceRecord.getServiceKey1())
-                .setServiceKey2(sourceRecord.getServiceKey2())
-                .setServiceKey3(sourceRecord.getServiceKey3())
-                .setServiceKey4(sourceRecord.getServiceKey4())
-                .setServiceKey5(sourceRecord.getServiceKey5());
+    private void fillCommonFields(TransferRecord tempRecord, Record originalRecord) {
```
Comment (RommeroInCountry) [inline src/main/java/com/incountry/residence/sdk/tools/transformer/EncryptedDtoTransformer.java:72]: let's better have have setRangeKeyFields and exclude setExpiresAt from it. 
setExpiresAt can go directly to getEncryptedTransferRecord

File: src/main/java/com/incountry/residence/sdk/tools/transformer/EncryptedDtoTransformer.java [hunk 9]
```diff
@@ -134,17 +139,7 @@ private TransferRecord getEncryptedTransferRecord(String complexBody, Record ori
...
-                .setRangeKey3(originalRecord.getRangeKey3())
-                .setRangeKey4(originalRecord.getRangeKey4())
-                .setRangeKey5(originalRecord.getRangeKey5())
-                .setRangeKey6(originalRecord.getRangeKey6())
-                .setRangeKey7(originalRecord.getRangeKey7())
-                .setRangeKey8(originalRecord.getRangeKey8())
-                .setRangeKey9(originalRecord.getRangeKey9())
-                .setRangeKey10(originalRecord.getRangeKey10())
-                .setExpiresAt(originalRecord.getExpiresAt());
+        fillCommonFields(resultRecord, originalRecord);
```
Comment (RommeroInCountry) [inline src/main/java/com/incountry/residence/sdk/tools/transformer/EncryptedDtoTransformer.java:142]: please have newlines in this methods dividing logical sections (setting profile/service keys/parent, setting bodies, setting ranges keys, setting search keys, setting date keys)

File: src/main/java/com/incountry/residence/sdk/tools/transformer/EncryptedDtoTransformer.java [hunk 10]
```diff
@@ -169,15 +164,6 @@ private TransferRecord getEncryptedTransferRecord(String complexBody, Record ori
         return resultRecord;
```
Comment (RommeroInCountry) [inline src/main/java/com/incountry/residence/sdk/tools/transformer/EncryptedDtoTransformer.java:164]: do we really need setAttachments(null) in this method?

File: src/main/java/com/incountry/residence/sdk/tools/transformer/EncryptedDtoTransformer.java [hunk 11]
```diff
@@ -1,30 +1,20 @@
-package com.incountry.residence.sdk.tools;
+package com.incountry.residence.sdk.tools.transformer;
```
Comment (RommeroInCountry) [inline src/main/java/com/incountry/residence/sdk/tools/transformer/EncryptedDtoTransformer.java:1]: this file has line
//readonly and number fields are copied
please update it to
// readonly and numeric fields are copied as is

**Review by RommeroInCountry — CHANGES_REQUESTED** [review]
